pkg/unpack/unpacker.go

```diff
@@ -47,48 +48,170 @@ const (
 	labelSnapshotRef = "containerd.io/snapshot.ref"
 )
 
+// Unpacker unpacks images by hooking into the image handler process.
+// Unpacks happen in the backgrounds and waited on to complete.
+type Unpacker interface {
+	Unpack(images.Handler) images.Handler
+	Wait() (Result, error)
+}
+
+// Result returns information about the unpacks which were completed.
+type Result struct {
+	Unpacks int
+}
+
+type unpackerConfig struct {
+	platforms []*Platform
+
+	content content.Store
+
+	limiter               *semaphore.Weighted
+	duplicationSuppressor kmutex.KeyedLocker
+}
+
+type Platform struct {
+	Platform platforms.Matcher
+
+	SnapshotterName string
+	Snapshotter     snapshots.Snapshotter
+	SnapshotOpts    []snapshots.Opt
+
+	Applier   diff.Applier
+	ApplyOpts []diff.ApplyOpt
+}
+
+type UnpackerOpt func(*unpackerConfig) error
+
+func WithUnpackPlatform(u Platform) UnpackerOpt {
+	return UnpackerOpt(func(c *unpackerConfig) error {
+		if u.Platform == nil {
+			u.Platform = platforms.All
+		}
+		if u.SnapshotterName == "" {
+			return fmt.Errorf("snapshotter name must be provided for reference")
+		}
+		if u.Snapshotter == nil {
+			return fmt.Errorf("snapshotter must be provided to unpack")
+		}
+		if u.Applier == nil {
+			return fmt.Errorf("applier must be provided to unpack")
+		}
+
+		c.platforms = append(c.platforms, &u)
+
+		return nil
+	})
+}
+
+func WithLimiter(l *semaphore.Weighted) UnpackerOpt {
+	return UnpackerOpt(func(c *unpackerConfig) error {
+		c.limiter = l
+		return nil
+	})
+}
+
+func WithDuplicationSuppressor(d kmutex.KeyedLocker) UnpackerOpt {
+	return UnpackerOpt(func(c *unpackerConfig) error {
+		c.duplicationSuppressor = d
+		return nil
+	})
+}
+
 type unpacker struct {
-	updateCh    chan ocispec.Descriptor
-	snapshotter string
-	config      UnpackConfig
-	c           *Client
-	limiter     *semaphore.Weighted
+	unpackerConfig
+
+	unpacks int32
+	ctx     context.Context
+	eg      *errgroup.Group
 }
 
-func (c *Client) newUnpacker(ctx context.Context, rCtx *RemoteContext) (*unpacker, error) {
-	snapshotter, err := c.resolveSnapshotterName(ctx, rCtx.Snapshotter)
-	if err != nil {
-		return nil, err
-	}
-	var config = UnpackConfig{
-		DuplicationSuppressor: kmutex.NewNoop(),
+func NewUnpacker(ctx context.Context, cs content.Store, opts ...UnpackerOpt) (Unpacker, error) {
```

Comment:
Can we have this return a real type instead of an interface?
The way this is, it tends to make it difficult to dig into the implementation when debugging/digging in from higher levels.

Reply:
Yeah, that makes sense here. The package isn't likely to have multiple implementation inside the package anyway. I could see the possibility of the unpack being proxied, but probably not even around that interface.